ksql-engine/src/main/java/io/confluent/ksql/util/HybridQueryIdGenerator.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.util;
+
+
+public class HybridQueryIdGenerator implements QueryIdGenerator {
+  private DefaultQueryIdGenerator defaultQueryIdGenerator;
+  private QueryIdGeneratorUsingOffset queryIdGeneratorUsingOffset;
+  private QueryIdGenerator primaryQueryIdGenerator;
+
+  public HybridQueryIdGenerator() {
+    this(0L);
+  }
+
+  public HybridQueryIdGenerator(final long initialValue) {
+    this.defaultQueryIdGenerator = new DefaultQueryIdGenerator(initialValue);
+    this.queryIdGeneratorUsingOffset = new QueryIdGeneratorUsingOffset();
+    this.primaryQueryIdGenerator = this.defaultQueryIdGenerator;
+
+  }
+
+  public void updateOffset(final long offset) {
+    this.primaryQueryIdGenerator = queryIdGeneratorUsingOffset;
+    queryIdGeneratorUsingOffset.updateOffset(offset);
+  }
+
+  @Override
+  public String getNextId() {
+    return String.valueOf(primaryQueryIdGenerator.getNextId());
+  }
+
+  @Override
+  public QueryIdGenerator createSandbox() {
+    if (primaryQueryIdGenerator instanceof DefaultQueryIdGenerator) {
```

Comment:
Sandboxed generators just need to be sequential, right?  They'll never have the mode flipped by anything.

So I think all you need is something like:

```java
public QueryIdGenerator createSandbox() {
   return new SequentialQueryIdGenerator(activeGenerator.peekNext());
}
```

Note, you'll need to add `peekNext` to the generator interface.